### PR TITLE
Run generators with / ro and /tmp mounted

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -3740,8 +3742,45 @@ static int build_generator_environment(Manager *m, char ***ret) {
...
+        };
+        int r;
+
+        r = build_generator_environment(m, &ge);
+        if (r < 0)
+                return log_error_errno(r, "Failed to build generator environment: %m");
+
+        /* We leave /run read-write because that's where the output goes to,
+         * and we leave /sys as-is, because our code checks whether it is read-only
+         * do detect containerized execution environments. */
```
Comment (yuwata) [inline src/core/manager.c]: "do detect" -> "to detect"

Comment (keszybz) [reply src/core/manager.c]: Updated. I also fixed a bug in logging found by the CI.

File: src/core/manager.c [hunk 2]
```diff
@@ -3762,31 +3801,14 @@ static int manager_run_generators(Manager *m) {
...
-                m->lookup_paths.generator_late,
-                NULL,
-        };
-
-        r = build_generator_environment(m, &ge);
-        if (r < 0) {
-                log_error_errno(r, "Failed to build generator environment: %m");
-                goto finish;
+        r = safe_fork("(sd-gens)",
+                      FORK_RESET_SIGNALS | FORK_LOG | FORK_WAIT | FORK_NEW_MOUNTNS | FORK_MOUNTNS_SLAVE,
```
Comment (yuwata) [inline src/core/manager.c]: I guess `FORK_NEW_MOUNTNS | FORK_MOUNTNS_SLAVE` cannot work for user sessions.

File: src/core/manager.c [hunk 3]
```diff
@@ -3740,8 +3742,45 @@ static int build_generator_environment(Manager *m, char ***ret) {
...
+                m->lookup_paths.generator_late,
+                NULL,
+        };
+        int r;
+
+        r = build_generator_environment(m, &ge);
+        if (r < 0)
+                return log_error_errno(r, "Failed to build generator environment: %m");
+
+        /* We leave /run read-write because that's where the output goes to,
```
Comment (poettering) [inline src/core/manager.c]: why the whole of /run/? wouldn't /run/systemd/ suffice?

File: src/core/manager.c [hunk 4]
```diff
@@ -3740,8 +3742,45 @@ static int build_generator_environment(Manager *m, char ***ret) {
...
+         * and we leave /sys as-is, because our code checks whether it is read-only
+         * to detect containerized execution environments. */
+        r = bind_remount_recursive("/", MS_RDONLY, MS_RDONLY, STRV_MAKE("/run", "/sys"));
+        if (r < 0)
+                log_warning_errno(r, "Read-only bind remount failed, ignoring: %m");
+
+        (void) mount_nofollow_verbose(LOG_WARNING,
+                                      "tmpfs", "/tmp", "tmpfs",
+                                      MS_NOSUID|MS_NODEV,
+                                      "mode=01777" TMPFS_LIMITS_RUN);
```
Comment (poettering) [inline src/core/manager.c]: doesn't matter, but overmounting /tmp sounds like something that sooner or later should probably end up as new flag for safe_fork(), i.e. FORK_PRIVATE_TMP or so. Sounds like something we could use at other places too

File: src/fstab-generator/fstab-generator.c [hunk 5]
```diff
@@ -671,12 +671,12 @@ static int parse_fstab(bool initrd) {
 
                 if (path_is_read_only_fs("/sys") > 0) {
                         if (streq(what, "sysfs")) {
-                                log_info("Running in a container, ignoring fstab entry for %s.", what);
+                                log_info("/sys is read-only, ignoring fstab entry for %s.", me->mnt_dir);
```
Comment (poettering) [inline src/fstab-generator/fstab-generator.c]: `/sys` → `/sys/`, please.

And I think we should keep the reference to the container, i.e. maybe reword to:

```
"/sys/ is read-only (running in a container?), ignoring fstab entry for '%s'."
```

or so?

File: src/core/manager.c [hunk 6]
```diff
@@ -3740,8 +3742,45 @@ static int build_generator_environment(Manager *m, char ***ret) {
...
+        int r;
+
+        r = build_generator_environment(m, &ge);
+        if (r < 0)
+                return log_error_errno(r, "Failed to build generator environment: %m");
+
+        /* We leave /run read-write because that's where the output goes to,
+         * and we leave /sys as-is, because our code checks whether it is read-only
+         * to detect containerized execution environments. */
+        r = bind_remount_recursive("/", MS_RDONLY, MS_RDONLY, STRV_MAKE("/run", "/sys"));
```
Comment (poettering) [inline src/core/manager.c]: /proc/ must remain writable too, btw, much of the Linux process API is exposed in procfs, including in writable files (for example selinux label stuff)

File: src/core/manager.c [hunk 7]
```diff
@@ -3740,8 +3742,45 @@ static int build_generator_environment(Manager *m, char ***ret) {
...
+        if (r < 0)
+                log_warning_errno(r, "Read-only bind remount failed, ignoring: %m");
```
Comment (poettering) [inline src/core/manager.c]: i have the suspicion some people will complain that they lose write access to /dev/shm/ which they uses as poor people's replacement for /tmp/ given that it is just a tmpfs and unlike /tmp/ definitely mounted at all times we invoke generators.

File: src/core/manager.c (hunk 3 above)
Comment (keszybz) [reply src/core/manager.c]: I don't think we should cut of /run, for the same reason that we leave /dev/shm: /run is expected to be writeable. Right now I can just leave /run by not remounting it, and to remount it ro and then punch holes is a lot more code.

**Comment by keszybz** [comment]
Updated. I added `MOUNT_PRIVATE_TMP` flag to safe_fork(), and it's now used by the code. The user manager skips the remounting, so tests should pass now.

File: src/core/manager.c [hunk 8]
```diff
@@ -3762,30 +3801,22 @@ static int manager_run_generators(Manager *m) {
...
-                m->lookup_paths.generator_early,
-                m->lookup_paths.generator_late,
-                NULL,
-        };
-
-        r = build_generator_environment(m, &ge);
-        if (r < 0) {
-                log_error_errno(r, "Failed to build generator environment: %m");
+        /* If we are the system manager, we fork and invoke the generators in a sanitized mount namespace. If
+         * we are the user manager, let's just execute the generators directly. We might now have the
```
Comment (yuwata) [inline src/core/manager.c]: "We might now have" -> "We might not have"

File: src/core/manager.c [hunk 9]
```diff
@@ -3762,30 +3801,22 @@ static int manager_run_generators(Manager *m) {
...
+         * necessary privileges, and the system manager has already mounted /tmp/ and everything else for us.
+         */
+        if (MANAGER_IS_USER(m)) {
+                r = manager_execute_generators(m, paths, false);
```
Comment (yuwata) [inline src/core/manager.c]: Missing usual short comment for the boolean argument.

File: src/core/manager.c [hunk 10]
```diff
@@ -3762,30 +3801,22 @@ static int manager_run_generators(Manager *m) {
...
-                                (char**) argv,
-                                ge,
-                                EXEC_DIR_PARALLEL | EXEC_DIR_IGNORE_ERRORS | EXEC_DIR_SET_SYSTEMD_EXEC_PID);
-
-        r = 0;
+        r = safe_fork("(sd-gens)",
+                      FORK_RESET_SIGNALS | FORK_LOG | FORK_WAIT | FORK_NEW_MOUNTNS | FORK_MOUNTNS_SLAVE | FORK_PRIVATE_TMP,
+                      NULL);
+        if (r == 0) {
+                r = manager_execute_generators(m, paths, true);
```
Comment (yuwata) [inline src/core/manager.c]: Also here.

**Comment by keszybz** [comment]
CentOS CI (Arch Linux) — TEST-75-RESOLVED, unrelated.

File: src/basic/process-util.h [hunk 11]
```diff
@@ -147,11 +147,12 @@ typedef enum ForkFlags {
         FORK_WAIT               = 1 <<  7, /* Wait until child exited */
         FORK_NEW_MOUNTNS        = 1 <<  8, /* Run child in its own mount namespace */
         FORK_MOUNTNS_SLAVE      = 1 <<  9, /* Make child's mount namespace MS_SLAVE */
-        FORK_RLIMIT_NOFILE_SAFE = 1 << 10, /* Set RLIMIT_NOFILE soft limit to 1K for select() compat */
-        FORK_STDOUT_TO_STDERR   = 1 << 11, /* Make stdout a copy of stderr */
-        FORK_FLUSH_STDIO        = 1 << 12, /* fflush() stdout (and stderr) before forking */
-        FORK_NEW_USERNS         = 1 << 13, /* Run child in its own user namespace */
-        FORK_CLOEXEC_OFF        = 1 << 14, /* In the child: turn off O_CLOEXEC on all fds in except_fds[] */
+        FORK_PRIVATE_TMP        = 1 << 10, /* Mount new /tmp/ in the child */
```
Comment (poettering) [inline src/basic/process-util.h]: maybe add a note to the comment "(Combine with FORK_NEWMOUNTNS|FORK_MOUNTNS_SLAVE!)"

if people don't do that, then the implications for the host will be weird...

Comment (yuwata) [reply src/basic/process-util.h]: Then, how about to explicitly request at least one of the flags?

**Comment by poettering** [comment]
lgtm (just the nitpick about the comment)

File: src/basic/process-util.h (hunk 11 above)
Comment (poettering) [reply src/basic/process-util.h]: hmm, you mean FORK_PRIVATE_TMP should imply the other two flags?

Comment (poettering) [inline src/basic/process-util.h]: the commit msg confusingly talks about MOUNT_PRIVATE_TMP rather than FORK_PRIVATE_TMP btw

Comment (yuwata) [reply src/basic/process-util.h]: I mean ignore the FORK_PRIVATE_TMP when at least one of FORK_NEWMOUNTNS and FORK_MOUNTNS_SLAVE is not set.

Comment (poettering) [reply src/basic/process-util.h]: yeah, that would probably make sense. either ignore, or maybe even assert() if they are missing

Comment (keszybz) [reply src/basic/process-util.h]: I added an assert on FORK_NEWMOUNTNS. I don't think we need the other flag: whether it's slave or not, doesn't really matter for this case.
